numpy/ma/core.py

```diff
@@ -3119,6 +3132,85 @@ def __array_wrap__(self, obj, context=None):
 
         return result
 
+    def __array_ufunc__(self, np_ufunc, method, *inputs, **kwargs):
+        """
+        MaskedArray capability for ufuncs
+
+        Handle masked versions of ufuncs if they are implemented within
+        the MaskedArray module. If the masked ufunc is not implemented,
+        this falls back to the standard numpy ndarray ufunc, which we
+        then call with the ndarray view of the input data.
+
+        """
+        # Output can be specified as arguments or as keyword arguments
+        outputs = kwargs.pop('out', ())
+        if not isinstance(outputs, tuple):
+            outputs = (outputs,)
+        outputs += inputs[np_ufunc.nin:]
+        args = inputs[:np_ufunc.nin]
+
+        # Determine what class types we are compatible with and return
+        # NotImplemented if we don't know how to handle them
+        for arg in args + outputs:
+            if not isinstance(arg, (ndarray, np.bool_, Number, list, str)):
```

Comment:
For `outputs`, one should be stricter: they need to be `ndarray` - indeed, for most functions, they should be `MaskedArray`, otherwise the mask cannot be stored.

Reply:
I think one could be a bit less strict on the inputs - after all, the unwrapped call will do its own checks anyway.